homeassistant/components/developer_credentials/__init__.py

```diff
@@ -0,0 +1,270 @@
+
+        await super().async_delete_item(item_id)
+
+
+class DeveloperCredentialsStorageListener:
+    """Listener that handles registering and unregistering OAuth implementations for credentials."""
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize DeveloperCredentialsStorageListener."""
+        self.hass = hass
+
+    async def updated(self, change_type: str, item_id: str, config: dict):
+        """Update set of registered authentication implementations."""
+        if change_type not in [collection.CHANGE_ADDED, collection.CHANGE_REMOVED]:
+            # not expected
+            return
+        integration_domain = config[CONF_DOMAIN]
+        developer_credential = DeveloperCredential(
+            config[CONF_CLIENT_ID], config[CONF_CLIENT_SECRET]
+        )
+        if change_type == collection.CHANGE_REMOVED:
+            _async_unregister_auth_implementation(
+                self.hass, integration_domain, developer_credential
+            )
+        else:
+            await _async_register_auth_implementation(
+                self.hass, integration_domain, developer_credential
+            )
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up Developer Credentials."""
+    hass.data[DOMAIN] = {}
+
+    await integration_platform.async_process_integration_platforms(
+        hass, DOMAIN, _register_developer_credentials_platform
+    )
+
+    # Developer Credentials from storage
+    id_manager = collection.IDManager()
+    storage_collection = DeveloperCredentialsStorageCollection(
+        Store(hass, STORAGE_VERSION, STORAGE_KEY),
+        logging.getLogger(f"{__name__}.storage_collection"),
+        id_manager,
+    )
+    storage_listener = DeveloperCredentialsStorageListener(hass)
+    storage_collection.async_add_listener(storage_listener.updated)
+    await storage_collection.async_load()
+
+    collection.StorageCollectionWebsocket(
+        storage_collection, DOMAIN, DOMAIN, CREATE_FIELDS, UPDATE_FIELDS
+    ).async_setup(hass)
+
+    # Allow future registration of local oauth implementations
+    config_entry_oauth2_flow.async_register_local_apis(hass)
+
+    return True
+
+
+# Creates AbstractOAuth2Implementation given a DeveloperCredential for a specific domain
+AuthImplFactory = Callable[
+    [str, DeveloperCredential], config_entry_oauth2_flow.AbstractOAuth2Implementation
+]
+
+
+def _get_auth_domain(domain: str, developer_credential: DeveloperCredential) -> str:
+    """Return the OAuth2 flow implementation domain."""
+    return slugify(f"{domain}.{developer_credential.client_id}")
+
+
+async def _async_register_auth_implementation(
+    hass: HomeAssistant,
+    domain: str,
+    developer_credential: DeveloperCredential,
+) -> None:
+    """Register an OAuth2 flow implementation for an integration."""
+    auth_domain = _get_auth_domain(domain, developer_credential)
+    authorization_server = await _async_get_authorization_server(hass, domain)
```

Comment:
Let's raise if already registered. Shouldn't happen so good to catch.
```suggestion
    if auth_domain in hass.data[DOMAIN][domain]:
        raise ValueError(f"Domain {auth_domain} already registered")
    authorization_server = await _async_get_authorization_server(hass, domain)
```